README.md
# DAPI Client

[![NPM Version](https://img.shields.io/npm/v/@dashevo/dapi-client)](https://www.npmjs.com/package/@dashevo/dapi-client)
[![Build Status](https://travis-ci.com/dashevo/js-dapi-client.svg?branch=master)](https://travis-ci.com/dashevo/js-dapi-client)
[![Release Date](https://img.shields.io/github/release-date/dashevo/js-dapi-client)](https://github.com/dashevo/js-dapi-client/releases/latest)
[![standard-readme compliant](https://img.shields.io/badge/readme%20style-standard-brightgreen)](https://github.com/RichardLitt/standard-readme)

Client library used to access Dash DAPI endpoints

This library enables HTTP-based interaction with the Dash blockchain and Dash
Platform via the decentralized API ([DAPI](https://github.com/dashevo/dapi))
hosted on Dash masternodes.

 - `DAPI-Client` provides automatic server (masternode) discovery using either a default seed node or a user-supplied one
 - `DAPI-Client` maps to DAPI's [RPC](https://github.com/dashevo/dapi/tree/master/lib/rpcServer/commands) and [gRPC](https://github.com/dashevo/dapi/tree/master/lib/grpcServer/handlers) endpoints

## Table of Contents
- [Install](#install)
- [Usage](#usage)
- [Documentation](#documentation)
- [Contributing](#contributing)
- [License](#license)

## Install

```sh
npm install @dashevo/dapi-client
```

## Usage

### Basic

```javascript
const DAPIClient = require('@dashevo/dapi-client');
const client = new DAPIClient();

client.core.getStatus().then((coreStatus) => {
  console.dir(coreStatus);
});
```

### Custom seed node

Custom seed nodes are necessary for connecting the client to devnets since the client library is unaware of them otherwise.

```javascript
const DAPIClient = require('@dashevo/dapi-client');

var client = new DAPIClient({
  seeds: [{
     host: 'seed-1.evonet.networks.dash.org',
     httpPort: 3000,
     grpcPort: 3010,
  }],
});

client.core.getBestBlockHash().then((r) => {
  console.log(r);
});
```

**Note**: The seed node shown above (`seed-1.evonet.networks.dash.org`) is for the Dash Evonet testing network.

### Custom addresses

Custom addresses may be directly specified in cases where it is beneficial to know exactly what node(s) are being accessed (e.g. debugging, local development, etc.).

```javascript
const DAPIClient = require('@dashevo/dapi-client');

var client = new DAPIClient({
  dapiAddresses: [
    '127.0.0.1:3000:3010',
    '127.0.0.2:3000:3010',
  ],
});

client.core.getBestBlockHash().then((r) => {
  console.log(r);
});
```

### Command specific options

DAPI Client options can be passed directly to any command to override any predefined client options and modify the client's behavior for that specific call.

```javascript
const DAPIClient = require('@dashevo/dapi-client');

// Set options to direct the request to a specific address and disable retries
const options = {
  dapiAddresses: ['127.0.0.1'],
  retries: 0,
};

client.core.getBestBlockHash(options).then((r) => {
  console.log(r);
});
```

## Documentation

More extensive documentation available at https://dashevo.github.io/dapi-client/.


## Contributing

Feel free to dive in! [Open an issue](https://github.com/dashevo/js-dapi-client/issues/new/choose) or submit PRs.

## License

[MIT](LICENSE) &copy; Dash Core Group, Inc.


CHANGELOG.md
## [0.21.3](https://github.com/dashevo/dapi-client/compare/v0.21.2...v0.21.3) (2021-11-12)


### Bug Fixes

* retry doesn’t work with 502 errors ([#291](https://github.com/dashevo/js-dapi-client/pull/291))


## [0.21.2](https://github.com/dashevo/dapi-client/compare/v0.21.1...v0.21.2) (2021-10-22)


### Features

* DAPI addresses white list must not be applied in Regtest mode ([#289](https://github.com/dashevo/js-dapi-client/pull/289))


## [0.21.1](https://github.com/dashevo/dapi-client/compare/v0.21.0...v0.21.1) (2021-10-21)


### Features

* DAPI addresses white list ([#288](https://github.com/dashevo/dapi-client/issues/288))



# [0.21.0](https://github.com/dashevo/dapi-client/compare/v0.20.0...v0.21.0) (2021-10-12)


### Features

* improve error handling ([#268](https://github.com/dashevo/dapi-client/issues/268))
* convenient response errors ([#272](https://github.com/dashevo/dapi-client/issues/272), [#280](https://github.com/dashevo/dapi-client/issues/280), [#282](https://github.com/dashevo/dapi-client/issues/282), [#284](https://github.com/dashevo/dapi-client/issues/284), [#286](https://github.com/dashevo/dapi-client/issues/286))
* implement getConsensusParams method ([#261](https://github.com/dashevo/dapi-client/issues/261))
* support returning of a multiproof ([#257](https://github.com/dashevo/dapi-client/issues/257))


### BREAKING CHANGES

* `getBlockByHash` and `getBlockByHeight` will throw `NotFoundError` instead of `null` result



## [0.20.1](https://github.com/dashevo/dapi-client/compare/v0.20.0...v0.20.1) (2021-05-26)


### Bug Fixes

* crash when connecting to an older version of dapi ([#252](https://github.com/dashevo/dapi-client/issues/252), [#253](https://github.com/dashevo/dapi-client/issues/253))



# [0.20.0](https://github.com/dashevo/dapi-client/compare/v0.19.3...v0.20.0) (2021-07-09)


### Features

* provide metadata in responses ([#243](https://github.com/dashevo/dapi-client/issues/243), [#246](https://github.com/dashevo/dapi-client/issues/246))
* throw `NotFoundError` if data is not found ([#247](https://github.com/dashevo/dapi-client/issues/247))


### BREAKING CHANGES

* platform methods respond with response classes instead of plain data
* Core method `getTransaction` responds with response class instead of plain data
* `getTransaction`, `getDataContract` and `getIdentity` methods throw `NotFoundError` if no data was found



## [0.19.3](https://github.com/dashevo/dapi-client/compare/v0.19.2...v0.19.3) (2021-05-26)


### Bug Fixes

* update cbor to fix serialization of queries with buffers ([178943b](https://github.com/dashevo/dapi-client/commit/178943b465b30ed838cb03a5a10c6f939db92a39))



## [0.19.2](https://github.com/dashevo/dapi-client/compare/v0.19.1...v0.19.2) (2021-05-20)


### Chores

* Update DPP to a 0.19.2 ([#237](https://github.com/dashevo/dapi-client/issues/237))



## [0.19.1](https://github.com/dashevo/dapi-client/compare/v0.19.0...v0.19.1) (2021-05-04)


### Bug Fixes

* `getStatus` reponse was returning buffers as Base64 encoded strings ([#235](https://github.com/dashevo/dapi-client/issues/235))



# [0.19.0](https://github.com/dashevo/dapi-client/compare/v0.18.0...v0.19.0) (2021-05-03)


### Features

* `getStatus` method handler response updated ([#229](https://github.com/dashevo/dapi-client/issues/229), [#230](https://github.com/dashevo/dapi-client/issues/230))


### Bug Fixes

* critical security vulnerability in axios@0.19.2 ([#233](https://github.com/dashevo/dapi-client/issues/233))


### BREAKING CHANGES

* `getStatus` method handler response have changed



# [0.18.0](https://github.com/dashevo/dapi-client/compare/v0.17.2...v0.18.0) (2021-03-03)


### Bug Fixes

* BLS was throwing an error inside `uncaughtException` handler ([#226](https://github.com/dashevo/dapi-client/issues/226))


### Features


* `waitForStateTransitionResult` method ([#221](https://github.com/dashevo/dapi-client/issues/221), ([#223](https://github.com/dashevo/dapi-client/issues/223))


### Chores

* remove temporary `timeout` option from `broadcastStateTransition` ([2abdf5](https://github.com/dashevo/js-dapi-client/commit/2abdf5dc45859bc142dd0f293d9f071190f2a59d))



## [0.17.2](https://github.com/dashevo/dapi-client/compare/v0.17.1...v0.17.2) (2020-12-30)


### Bug Fixes

* broadcastStateTransitions is timing out on testnet ([#219](https://github.com/dashevo/dapi-client/issues/219))



## [0.17.1](https://github.com/dashevo/dapi-client/compare/v0.17.0...v0.17.1) (2020-12-30)


### Bug Fixes

* merkleRootQuorums from the diff doesn’t match calculated quorum root after diff is applied ([#217](https://github.com/dashevo/dapi-client/issues/217))



# [0.17.0](https://github.com/dashevo/dapi-client/compare/v0.16.0...v0.17.0) (2020-12-29)


### Features

* introduce testnet network ([#214](https://github.com/dashevo/dapi-client/issues/214))
* update `dpp` and `dashcore-lib` ([#207](https://github.com/dashevo/dapi-client/issues/207), [#210](https://github.com/dashevo/dapi-client/issues/210), [#211](https://github.com/dashevo/dapi-client/issues/211), [#212](https://github.com/dashevo/dapi-client/issues/212))


### Bug Fixes

* SML unhandled error on blockchain reorg ([#215](https://github.com/dashevo/dapi-client/issues/215))


### BREAKING CHANGES

* DAPI client is now connecting to a testnet by default



# [0.16.0](https://github.com/dashevo/dapi-client/compare/v0.15.0...v0.16.0) (2020-10-27)


### Features

* `getIdentitiesByPublicKeyHashes` and `getIdentityIdsByPublicKeyHashes` methods ([#191](https://github.com/dashevo/dapi-client/issues/191), [#196](https://github.com/dashevo/dapi-client/issues/196), [#205](https://github.com/dashevo/dapi-client/issues/205))
* `getDataContract`, `getDocuments`, `getIdentity` accept `Buffer` ([#201](https://github.com/dashevo/dapi-client/issues/201))


### Documentation

* fix URLs in README ([#193](https://github.com/dashevo/dapi-client/issues/193))


### BREAKING CHANGES


src/index.js
const jsutil = require('@dashevo/dashcore-lib').util.js;
const preconditionsUtil = require('@dashevo/dashcore-lib').util.preconditions;
const cbor = require('cbor');

const {
  TransactionsWithProofsRequest,
  BloomFilter: BloomFilterMessage,
  ApplyStateTransitionRequest,
  GetIdentityRequest,
  GetDataContractRequest,
  GetDocumentsRequest,
  GetBlockRequest,
  GetStatusRequest,
  GetTransactionRequest,
  SendTransactionRequest,
  GetIdentityByFirstPublicKeyRequest,
  GetIdentityIdByFirstPublicKeyRequest,
} = require('@dashevo/dapi-grpc');

const ListMNDiscovery = require('./MNDiscovery/ListMNDiscovery');
const JsonRpcTransport = require('./transport/JsonRpcTransport');
const MasternodeListProvider = require('./MNDiscovery/MasternodeListProvider');

const MNDiscovery = require('./MNDiscovery/index');
const TransportManager = require('./transport/TransportManager');
const config = require('./config');
const { responseErrorCodes } = require('./constants');

/**
 * @param options
 * @param {Array<object>} [options.seeds] - seeds. If no seeds provided
 * default seed will be used.
 * @param {number} [options.port=3000] - default port for connection to the DAPI
 * @param {number} [options.nativeGrpcPort=3010] - Native GRPC port for connection to the DAPI
 * @param {number} [options.timeout=2000] - timeout for connection to the DAPI
 * @param {number} [options.retries=3] - num of retries if there is no response from DAPI node
 */
class DAPIClient {
  constructor(options = {}) {
    this.DAPIPort = options.port || config.Api.port;

    this.nativeGrpcPort = options.nativeGrpcPort || config.grpc.nativePort;

    this.timeout = options.timeout || 2000;
    preconditionsUtil.checkArgument(jsutil.isUnsignedInteger(this.timeout),
      'Expect timeout to be an unsigned integer');

    this.retries = options.retries ? options.retries : 3;
    preconditionsUtil.checkArgument(jsutil.isUnsignedInteger(this.retries),
      'Expect retries to be an unsigned integer');

    const listMNDiscovery = new ListMNDiscovery(options.seeds);

    const jsonRpcTransport = new JsonRpcTransport(
      listMNDiscovery, this.DAPIPort,
    );

    const masternodeListProvider = new MasternodeListProvider(jsonRpcTransport);

    this.MNDiscovery = new MNDiscovery(masternodeListProvider);

    this.transportManager = new TransportManager(
      this.MNDiscovery, this.DAPIPort, this.nativeGrpcPort,
    );
  }

  /* Layer 1 commands */
  /**
   * ONLY FOR TESTING PURPOSES WITH REGTEST. WILL NOT WORK ON TESTNET/LIVENET.
   *
   * @param {number} blocksNumber - Number of blocks to generate
   * @param {string} address - The address that will receive the newly generated Dash
   * @returns {Promise<string[]>} - block hashes
   */
  generateToAddress(blocksNumber, address) {
    return this.transportManager.get(TransportManager.JSON_RPC)
      .makeRequest(
        'generateToAddress', { blocksNumber, address },
        { retriesCount: this.retries, client: { timeout: this.timeout } },
      );
  }

  /**
   * Returns block hash of chaintip
   *
   * @returns {Promise<string>}
   */
  getBestBlockHash() {
    return this.transportManager.get(TransportManager.JSON_RPC)
      .makeRequest(
        'getBestBlockHash', {},
        { retriesCount: this.retries, client: { timeout: this.timeout } },
      );
  }

  /**
   * Returns block hash for the given height
   *
   * @param {number} height
   * @returns {Promise<string>} - block hash
   */
  getBlockHash(height) {
    return this.transportManager.get(TransportManager.JSON_RPC)
      .makeRequest(
        'getBlockHash', { height },
        { retriesCount: this.retries, client: { timeout: this.timeout } },
      );
  }

  /**
   * Get deterministic masternodelist diff
   *
   * @param {string} baseBlockHash - hash or height of start block
   * @param {string} blockHash - hash or height of end block
   * @returns {Promise<object>}
   */
  getMnListDiff(baseBlockHash, blockHash) {
    return this.transportManager.get(TransportManager.JSON_RPC)
      .makeRequest(
        'getMnListDiff', { baseBlockHash, blockHash },
        { retriesCount: this.retries, client: { timeout: this.timeout } },
      );
  }

  /**
   * Returns a summary (balance, txs) for a given address
   *
   * @param {string|string[]} address or array of addresses
   * @param {boolean} [noTxList=false] - true if a list of all txs should NOT be included in result
   * @param {number} [from] - start of range for the tx to be included in the tx list
   * @param {number} [to] - end of range for the tx to be included in the tx list
   * @param {number} [fromHeight] - which height to start from (optional, overriding from/to)
   * @param {number} [toHeight] - on which height to end (optional, overriding from/to)
   * @returns {Promise<object>} - an object with basic address info
   */
  getAddressSummary(address, noTxList, from, to, fromHeight, toHeight) {
    return this.transportManager.get(TransportManager.JSON_RPC).makeRequest(
      'getAddressSummary',
      {
        address, noTxList, from, to, fromHeight, toHeight,
      },
      { retriesCount: this.retries, client: { timeout: this.timeout } },
    );
  }

  /**
   * Get block by height
   *
   * @param {number} height
   * @returns {Promise<null|Buffer>}
   */
  async getBlockByHeight(height) {
    const getBlockRequest = new GetBlockRequest();
    getBlockRequest.setHeight(height);

    let response;
    try {
      response = await this.transportManager.get(TransportManager.GRPC_CORE)
        .makeRequest(
          'getBlock', getBlockRequest,
          { retriesCount: this.retries, client: { timeout: this.timeout } },
        );
    } catch (e) {
      if (e.code === responseErrorCodes.NOT_FOUND) {
        return null;
      }

      throw e;
    }

    const blockBinaryArray = response.getBlock();

    return Buffer.from(blockBinaryArray);
  }

  /**
   * Get block by hash
   *
   * @param {string} hash
   * @returns {Promise<null|Buffer>}
   */
  async getBlockByHash(hash) {
    const getBlockRequest = new GetBlockRequest();
    getBlockRequest.setHash(hash);

    let response;
    try {
      response = await this.transportManager.get(TransportManager.GRPC_CORE)
        .makeRequest(
          'getBlock', getBlockRequest,
          { retriesCount: this.retries, client: { timeout: this.timeout } },
        );
    } catch (e) {
      if (e.code === responseErrorCodes.NOT_FOUND) {
        return null;
      }

      throw e;
    }

    const blockBinaryArray = response.getBlock();

    return Buffer.from(blockBinaryArray);
  }

  /**
   * Get Core chain status
   *
   * @returns {Promise<object>}
   */
  async getStatus() {
    const getStatusRequest = new GetStatusRequest();

    const response = await this.transportManager.get(TransportManager.GRPC_CORE)
      .makeRequest(
        'getStatus', getStatusRequest,
        { retriesCount: this.retries, client: { timeout: this.timeout } },
      );

    return response.toObject();
  }

  /**
   * Get Transaction by ID
   *
   * @param {string} id
   * @returns {Promise<null|Buffer>}
   */
  async getTransaction(id) {
    const getTransactionRequest = new GetTransactionRequest();
    getTransactionRequest.setId(id);

    let response;
    try {
      response = await this.transportManager.get(TransportManager.GRPC_CORE)
        .makeRequest(
          'getTransaction', getTransactionRequest,
          { retriesCount: this.retries, client: { timeout: this.timeout } },
        );
    } catch (e) {
      if (e.code === responseErrorCodes.NOT_FOUND) {
        return null;
      }

      throw e;
    }

    const transactionBinaryArray = response.getTransaction();

    let transaction = null;
    if (transactionBinaryArray) {
      transaction = Buffer.from(transactionBinaryArray);
    }

    return transaction;
  }

  /**
   * Send Transaction
   *
   * @param {Buffer} transaction
   * @param {object} [options]
   * @param {object} [options.allowHighFees=false]
   * @param {object} [options.bypassLimits=false]
   * @returns {Promise<string>}
   */
  async sendTransaction(transaction, options = {}) {
    const sendTransactionRequest = new SendTransactionRequest();
    sendTransactionRequest.setTransaction(transaction);
    sendTransactionRequest.setAllowHighFees(options.allowHighFees || false);
    sendTransactionRequest.setBypassLimits(options.bypassLimits || false);

    const response = await this.transportManager.get(TransportManager.GRPC_CORE)
      .makeRequest(
        'sendTransaction', sendTransactionRequest,
        { retriesCount: this.retries, client: { timeout: this.timeout } },
      );

    return response.getTransactionId();
  }

  /**
   * Returns UTXO for a given address or multiple addresses (max result 1000)
   *
   * @param {string|string[]} address or array of addresses
   * @param {number} [from] - start of range in the ordered list of latest UTXO (optional)
   * @param {number} [to] - end of range in the ordered list of latest UTXO (optional)
   * @param {number} [fromHeight] - which height to start from (optional, overriding from/to)
   * @param {number} [toHeight] - on which height to end (optional, overriding from/to)
   * @returns {Promise<object>} - Object with pagination info and array of unspent outputs
   */
  getUTXO(address, from, to, fromHeight, toHeight) {
    return this.transportManager.get(TransportManager.JSON_RPC).makeRequest(
      'getUTXO',
      {
        address, from, to, fromHeight, toHeight,
      },
      { retriesCount: this.retries, client: { timeout: this.timeout } },
    );
  }

  /* gRPC methods */

  /* txFilterStream methods */
  /**
   * @param {object} bloomFilter
   * @param {Uint8Array|Array} bloomFilter.vData - The filter itself is simply a bit
   * field of arbitrary byte-aligned size. The maximum size is 36,000 bytes.
   * @param {number} bloomFilter.nHashFuncs - The number of hash functions to use in this filter.
   * The maximum value allowed in this field is 50.
   * @param {number} bloomFilter.nTweak - A random value to add to the seed value in the
   * hash function used by the bloom filter.
   * @param {number} bloomFilter.nFlags - A set of flags that control how matched items
   * are added to the filter.
   * @param {object} [options]
   * @param {string} [options.fromBlockHash] - Specifies block hash to start syncing from
   * @param {number} [options.fromBlockHeight] - Specifies block height to start syncing from
   * @param {number} [options.count=0] - Number of blocks to sync,
   * if set to 0 syncing is continuously sends new data as well
   * @returns {
   *    Promise<EventEmitter>|!grpc.web.ClientReadableStream<!TransactionsWithProofsResponse>
   * }
   */
  async subscribeToTransactionsWithProofs(bloomFilter, options = { count: 0 }) {
    const bloomFilterMessage = new BloomFilterMessage();

    let { vData } = bloomFilter;
    if (Array.isArray(vData)) {
      vData = new Uint8Array(vData);
    }

    bloomFilterMessage.setVData(vData);
    bloomFilterMessage.setNHashFuncs(bloomFilter.nHashFuncs);
    bloomFilterMessage.setNTweak(bloomFilter.nTweak);
    bloomFilterMessage.setNFlags(bloomFilter.nFlags);

    const request = new TransactionsWithProofsRequest();
    request.setBloomFilter(bloomFilterMessage);

    if (options.fromBlockHeight) {
      request.setFromBlockHeight(options.fromBlockHeight);
    }

    if (options.fromBlockHash) {
      request.setFromBlockHash(
        Buffer.from(options.fromBlockHash, 'hex'),
      );
    }

    request.setCount(options.count);

    return this.transportManager.get(TransportManager.GRPC_TX)
      .makeRequest(
        'subscribeToTransactionsWithProofs', request,
        { retriesCount: this.retries, client: { timeout: this.timeout } },
      );
  }

  /* Platform gRPC methods */

  /**
   * Send State Transition to machine
   *
   * @param {AbstractStateTransition} stateTransition
   * @returns {Promise<!ApplyStateTransitionResponse>}
   */
  async applyStateTransition(stateTransition) {
    const applyStateTransitionRequest = new ApplyStateTransitionRequest();
    applyStateTransitionRequest.setStateTransition(stateTransition.serialize());

    return this.transportManager.get(TransportManager.GRPC_PLATFORM)
      .makeRequest(
        'applyStateTransition', applyStateTransitionRequest,
        { retriesCount: this.retries, client: { timeout: this.timeout } },
      );
  }

  /**
   * Fetch the identity by id
   *
   * @param {string} id
   * @returns {Promise<!Buffer|null>}
   */
  async getIdentity(id) {
    const getIdentityRequest = new GetIdentityRequest();
    getIdentityRequest.setId(id);

    let getIdentityResponse;
    try {
      getIdentityResponse = await this.transportManager.get(TransportManager.GRPC_PLATFORM)
        .makeRequest(
          'getIdentity', getIdentityRequest,
          { retriesCount: this.retries, client: { timeout: this.timeout } },
        );
    } catch (e) {
      if (e.code === responseErrorCodes.NOT_FOUND) {
        return null;
      }

      throw e;
    }

    const serializedIdentityBinaryArray = getIdentityResponse.getIdentity();
    let identity = null;

    if (serializedIdentityBinaryArray) {
      identity = Buffer.from(serializedIdentityBinaryArray);
    }

    return identity;
  }

  /**
   * Fetch Data Contract by id
   *
   * @param {string} contractId
   * @returns {Promise<Buffer>}
   */
  async getDataContract(contractId) {
    const getDataContractRequest = new GetDataContractRequest();

    getDataContractRequest.setId(contractId);

    let getDataContractResponse;
    try {
      getDataContractResponse = await this.transportManager.get(TransportManager.GRPC_PLATFORM)
        .makeRequest(
          'getDataContract', getDataContractRequest,
          { retriesCount: this.retries, client: { timeout: this.timeout } },
        );
    } catch (e) {
      if (e.code === responseErrorCodes.NOT_FOUND) {
        return null;
      }

      throw e;
    }

    const serializedDataContractBinaryArray = getDataContractResponse.getDataContract();

    let dataContract = null;

    if (serializedDataContractBinaryArray) {
      dataContract = Buffer.from(serializedDataContractBinaryArray);
    }

    return dataContract;
  }

  /**
   * Fetch Documents from Drive
   *
   * @param {string} contractId
   * @param {string} type - Dap objects type to fetch
   * @param {object} options
   * @param {object} options.where - Mongo-like query
   * @param {object} options.orderBy - Mongo-like sort field
   * @param {number} options.limit - how many objects to fetch
   * @param {number} options.startAt - number of objects to skip
   * @param {number} options.startAfter - exclusive skip
   * @returns {Promise<Buffer[]>}
   */
  async getDocuments(contractId, type, options) {
    const {
      where,
      orderBy,
      limit,
      startAt,
      startAfter,
    } = options;

    let whereSerialized;
    if (where) {
      whereSerialized = cbor.encode(where);
    }

    let orderBySerialized;
    if (orderBy) {
      orderBySerialized = cbor.encode(orderBy);
    }

    const getDocumentsRequest = new GetDocumentsRequest();
    getDocumentsRequest.setDataContractId(contractId);
    getDocumentsRequest.setDocumentType(type);
    getDocumentsRequest.setWhere(whereSerialized);
    getDocumentsRequest.setOrderBy(orderBySerialized);
    getDocumentsRequest.setLimit(limit);
    getDocumentsRequest.setStartAfter(startAfter);
    getDocumentsRequest.setStartAt(startAt);

    const getDocumentsResponse = await this.transportManager.get(TransportManager.GRPC_PLATFORM)
      .makeRequest(
        'getDocuments', getDocumentsRequest,
        { retriesCount: this.retries, client: { timeout: this.timeout } },
      );

    return getDocumentsResponse.getDocumentsList().map((document) => Buffer.from(document));
  }

  /**
   * Fetch the identity by public key hash
   * @param {string} publicKeyHash
   * @return {Promise<!Buffer|null>}
   */
  async getIdentityByFirstPublicKey(publicKeyHash) {
    const getIdentityByFirstPublicKeyRequest = new GetIdentityByFirstPublicKeyRequest();
    getIdentityByFirstPublicKeyRequest.setPublicKeyHash(Buffer.from(publicKeyHash, 'hex'));

    let getIdentityByFirstPublicKeyResponse;
    try {
      getIdentityByFirstPublicKeyResponse = await this.transportManager
        .get(TransportManager.GRPC_PLATFORM)
        .makeRequest(
          'getIdentityByFirstPublicKey', getIdentityByFirstPublicKeyRequest,
          { retriesCount: this.retries, client: { timeout: this.timeout } },
        );
    } catch (e) {
      if (e.code === responseErrorCodes.NOT_FOUND) {
        return null;
      }

      throw e;
    }

    const serializedIdentityBinaryArray = getIdentityByFirstPublicKeyResponse.getIdentity();
    let identity = null;

    if (serializedIdentityBinaryArray) {
      identity = Buffer.from(serializedIdentityBinaryArray);
    }

    return identity;
  }

  /**
   * Fetch the identity id by public key hash
   * @param {string} publicKeyHash
   * @return {Promise<!string|null>}
   */
  async getIdentityIdByFirstPublicKey(publicKeyHash) {
    const getIdentityIdByFirstPublicKeyRequest = new GetIdentityIdByFirstPublicKeyRequest();
    getIdentityIdByFirstPublicKeyRequest.setPublicKeyHash(Buffer.from(publicKeyHash, 'hex'));

    let getIdentityIdByFirstPublicKeyResponse;
    try {
      getIdentityIdByFirstPublicKeyResponse = await this.transportManager
        .get(TransportManager.GRPC_PLATFORM)
        .makeRequest(
          'getIdentityIdByFirstPublicKey', getIdentityIdByFirstPublicKeyRequest,
          { retriesCount: this.retries, client: { timeout: this.timeout } },
        );
    } catch (e) {
      if (e.code === responseErrorCodes.NOT_FOUND) {
        return null;
      }

      throw e;
    }

    return getIdentityIdByFirstPublicKeyResponse.getId();
  }
}

module.exports = DAPIClient;
